ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -0,0 +1,302 @@
+
+          try {
+            reportingUrl = new URL(urlObj.getProtocol(), urlObj.getHost(),
+                urlObj.getPath() + "?" + queryString).toString();
+          } catch (MalformedURLException e) {
+            throw new InvalidUrlException(
+                "Could not parse reporting with query string: " + queryString, e);
+          }
+
+          attributes.put(Attribute.REPORTING_URL, reportingUrl);
+          payload.put(Attribute.REPORTING_URL, reportingUrl);
+
+          return new PubsubMessage(Json.asBytes(payload), attributes);
+        }).exceptionsInto(TypeDescriptor.of(PubsubMessage.class))
+            .exceptionsVia((ExceptionElement<PubsubMessage> ee) -> {
+              try {
+                throw ee.exception();
+              } catch (UncheckedIOException | IllegalArgumentException | InvalidUrlException e) {
+                return FailureMessage.of(ParseReportingUrl.class.getSimpleName(), ee.element(),
+                    ee.exception());
+              }
+            }));
+  }
+
+  @VisibleForTesting
+  boolean isUrlValid(URL url, String doctype) {
+    Set<String> allowedUrls;
+
+    if (doctype.endsWith("-click")) {
+      allowedUrls = singletonAllowedClickUrls;
+    } else if (doctype.endsWith("-impression")) {
+      allowedUrls = singletonAllowedImpressionUrls;
+    } else {
+      throw new IllegalArgumentException("Invalid doctype: " + doctype);
+    }
+
+    if (allowedUrls.contains(url.getHost())) {
+      return true;
+    }
+
+    // check for subdomains (e.g. allow mozilla.test.com but not mozillatest.com)
+    return allowedUrls.stream().map(allowedUrl -> "." + allowedUrl)
+        .anyMatch(url.getHost()::endsWith);
+  }
+
+  @VisibleForTesting
+  String createIpParam(String countryCode) {
+    String ipParam = singletonCountryToIpMapping.getOrDefault(countryCode,
+        singletonCountryToIpMapping.get(DEFAULT_COUNTRY));
+    if (ipParam == null) {
+      throw new IllegalArgumentException("Could not get ip value: Unrecognized country "
+          + "and missing default value: " + Attribute.NORMALIZED_COUNTRY_CODE);
+    }
+    return ipParam;
+  }
+
+  @VisibleForTesting
+  String createUserAgentParam(String os, String clientVersion) throws UnsupportedEncodingException {
+    String normalizedOs;
+    if (os.startsWith(OS_WINDOWS)) {
+      normalizedOs = OS_WINDOWS;
+    } else if (os.startsWith(OS_MAC)) {
+      normalizedOs = OS_MAC;
+    } else if (os.startsWith(OS_LINUX)) {
+      normalizedOs = OS_LINUX;
+    } else {
+      normalizedOs = DEFAULT_OS;
+    }
+    String userAgentFormatString = singletonOsToUserAgentMapping.getOrDefault(normalizedOs,
+        singletonOsToUserAgentMapping.get(DEFAULT_COUNTRY));
+    if (userAgentFormatString == null) {
+      throw new IllegalArgumentException(
+          "Could not get user agent value: Unrecognized OS and missing default value: " + os);
+    }
+    String userAgent = MessageFormat.format(userAgentFormatString, clientVersion);
```

Comment:
I have never interacted with `java.text.MessageFormat` and did not know it existed. Interesting.

Elsewhere in this codebase we use [commons-text StringSubstitutor](https://commons.apache.org/proper/commons-text/apidocs/org/apache/commons/text/StringSubstitutor.html) and I'd have a preference for using that here as well, in which case we'd replace the `{0}` placeholder with `{ua_os_version}` in the templates, which would be more descriptive and more extensible if the UA formatting becomes more complex in the future.